reference/ccr/apis/get-ccr-stats/line_94 docs test failure

```
java.lang.AssertionError: 
Failure at [reference/ccr/apis/get-ccr-stats:75]: $body didn't match expected value:
                         $body: 
               auto_follow_stats: 
            auto_followed_clusters: 
                                   0: 
                          cluster_name: same [remote_cluster]
            last_seen_metadata_version: same [11613]
          time_since_last_check_millis: same [0]
    number_of_failed_follow_indices: same [0]
    number_of_failed_remote_cluster_state_requests: same [3]
    number_of_successful_follow_indices: same [3]
         recent_auto_follow_errors: 
                                   0: 
                 auto_follow_exception: 
                               caused_by: 
                                 caused_by: 
                                   caused_by: 
                                        reason: same [[node-0][127.0.0.1:36317][cluster:monitor/xpack/in...]
                                          type: same [node_disconnected_exception]
                                      reason: same [could not determine the license type for cluster [...]
                                        type: same [exception]
                                    reason: same [can not fetch remote cluster state as the license ...]
                                      type: same [status_exception]
                                  reason: same [ElasticsearchStatusException[can not fetch remote ...]
                                    type: same [exception]
                          leader_index: same [my_auto_follow_pattern]
                             timestamp: same [1572451902549]
                    follow_stats: 
                           indices: 
                                   0: 
                                 index: same [follower_index]
                                shards: 
                                       0: 
                                bytes_read: same [0]
                      failed_read_requests: same [0]
                     failed_write_requests: same [0]
                  follower_aliases_version: same [0]
                follower_global_checkpoint: same [-1]
                            follower_index: same [follower_index]
                  follower_mapping_version: same [0]
                       follower_max_seq_no: same [-1]
                 follower_settings_version: same [0]
                     last_requested_seq_no: same [-1]
                  leader_global_checkpoint: same [-1]
                              leader_index: same [leader_index]
                         leader_max_seq_no: same [-1]
                           operations_read: same [0]
                        operations_written: same [0]
                 outstanding_read_requests: same [0]
                outstanding_write_requests: same [0]
                           read_exceptions: same [empty list]
                            remote_cluster: same [remote_cluster]
                                  shard_id: same [0]
                  successful_read_requests: same [0]
                 successful_write_requests: same [0]
               time_since_last_read_millis: same [-1]
            total_read_remote_exec_time_millis: same [0]
                    total_read_time_millis: same [0]
                   total_write_time_millis: same [0]
              write_buffer_operation_count: same [0]
                write_buffer_size_in_bytes: same [0]
      received [1] more entries than expected
```

See last line (^).

Build url: https://gradle-enterprise.elastic.co/s/6vk2bsh6vsdem/tests/jbm2x6tnwtxku-etwxtlodvtjas

This test failed [3 times](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'test:"test%20%7Byaml%3Dreference%2Fccr%2Fapis%2Fget-ccr-stats%2Fline_94%7D"'),sort:!(time,desc))) already today.

--- a/docs/reference/ccr/apis/auto-follow/pause-auto-follow-pattern.asciidoc
+++ b/docs/reference/ccr/apis/auto-follow/pause-auto-follow-pattern.asciidoc
@@ -63,6 +63,13 @@ PUT /_ccr/auto_follow/my_auto_follow_pattern
 // TEST[setup:remote_cluster]
 // TESTSETUP
 
+[source,console]
+--------------------------------------------------
+DELETE /_ccr/auto_follow/my_auto_follow_pattern
+--------------------------------------------------
+// TEST
+// TEARDOWN
+
 //////////////////////////
 
 [source,console]

--- a/docs/reference/ccr/apis/auto-follow/resume-auto-follow-pattern.asciidoc
+++ b/docs/reference/ccr/apis/auto-follow/resume-auto-follow-pattern.asciidoc
@@ -61,6 +61,13 @@ PUT /_ccr/auto_follow/my_auto_follow_pattern
 
 [source,console]
 --------------------------------------------------
+DELETE /_ccr/auto_follow/my_auto_follow_pattern
+--------------------------------------------------
+// TEST
+// TEARDOWN
+
+[source,console]
+--------------------------------------------------
 POST /_ccr/auto_follow/my_auto_follow_pattern/pause
 --------------------------------------------------
 // TEST

--- a/test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java
@@ -470,6 +470,15 @@ public abstract class ESRestTestCase extends ESTestCase {
     }
 
     /**
+     * Returns whether to preserve auto-follow patterns. Defaults to not
+     * preserving them. Only runs at all if xpack is installed on the cluster
+     * being tested.
+     */
+    protected boolean preserveAutoFollowPatternsUponCompletion() {
+        return false;
+    }
+
+    /**
      * Returns whether to preserve SLM Policies of this test. Defaults to not
      * preserving them. Only runs at all if xpack is installed on the cluster
      * being tested.
@@ -558,6 +567,10 @@ public abstract class ESRestTestCase extends ESTestCase {
 
         if (hasXPack && false == preserveILMPoliciesUponCompletion()) {
             deleteAllILMPolicies();
+        }
+
+        if (hasXPack && false == preserveAutoFollowPatternsUponCompletion()) {
+            deleteAllAutoFollowPatterns();
         }
 
         assertThat("Found in progress snapshots [" + inProgressSnapshots.get() + "].", inProgressSnapshots.get(), anEmptyMap());
@@ -733,6 +746,31 @@ public abstract class ESRestTestCase extends ESTestCase {
 
         for (String policyName : policies.keySet()) {
             adminClient().performRequest(new Request("DELETE", "/_slm/policy/" + policyName));
+        }
+    }
+
+    private static void deleteAllAutoFollowPatterns() throws IOException {
+        final List<Map<?, ?>> patterns;
+
+        try {
+            Response response = adminClient().performRequest(new Request("GET", "/_ccr/auto_follow"));
+            patterns = (List<Map<?, ?>>) entityAsMap(response).get("patterns");
+        } catch (ResponseException e) {
+            if (RestStatus.METHOD_NOT_ALLOWED.getStatus() == e.getResponse().getStatusLine().getStatusCode() ||
+                RestStatus.BAD_REQUEST.getStatus() == e.getResponse().getStatusLine().getStatusCode()) {
+                // If bad request returned, CCR is not enabled.
+                return;
+            }
+            throw e;
+        }
+
+        if (patterns == null || patterns.isEmpty()) {
+            return;
+        }
+
+        for (Map<?, ?> pattern : patterns) {
+            String patternName = (String) pattern.get("name");
+            adminClient().performRequest(new Request("DELETE", "/_ccr/auto_follow/" + patternName));
         }
     }
 

--- a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportFollowStatsAction.java
+++ b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportFollowStatsAction.java
@@ -13,9 +13,11 @@ import org.elasticsearch.action.TaskOperationFailure;
 import org.elasticsearch.action.support.ActionFilters;
 import org.elasticsearch.action.support.tasks.TransportTasksAction;
 import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.MetaData;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.index.Index;
 import org.elasticsearch.license.LicenseUtils;
 import org.elasticsearch.persistent.PersistentTasksCustomMetaData;
 import org.elasticsearch.tasks.Task;
@@ -116,15 +118,17 @@ public class TransportFollowStatsAction extends TransportTasksAction<
         if (persistentTasksMetaData == null) {
             return Collections.emptySet();
         }
-
+        final MetaData metaData = state.metaData();
         final Set<String> requestedFollowerIndices = indices != null ?
             new HashSet<>(Arrays.asList(indices)) : Collections.emptySet();
         return persistentTasksMetaData.tasks().stream()
             .filter(persistentTask -> persistentTask.getTaskName().equals(ShardFollowTask.NAME))
             .map(persistentTask -> {
                 ShardFollowTask shardFollowTask = (ShardFollowTask) persistentTask.getParams();
-                return shardFollowTask.getFollowShardId().getIndexName();
+                return shardFollowTask.getFollowShardId().getIndex();
             })
+            .filter(followerIndex -> metaData.index(followerIndex) != null) // hide tasks that are orphaned (see ShardFollowTaskCleaner)
+            .map(Index::getName)
             .filter(followerIndex -> Strings.isAllOrWildcard(indices) || requestedFollowerIndices.contains(followerIndex))
             .collect(Collectors.toSet());
     }

--- a/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/TransportFollowInfoActionTests.java
+++ b/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/TransportFollowInfoActionTests.java
@@ -11,6 +11,7 @@ import org.elasticsearch.cluster.ClusterName;
 import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.index.Index;
 import org.elasticsearch.persistent.PersistentTasksCustomMetaData;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.xpack.ccr.Ccr;
@@ -58,7 +59,8 @@ public class TransportFollowInfoActionTests extends ESTestCase {
             if (isFollowIndex) {
                 imdBuilder.putCustom(Ccr.CCR_CUSTOM_METADATA_KEY, new HashMap<>());
                 if (active) {
-                    persistentTasks.addTask(Integer.toString(i), ShardFollowTask.NAME, createShardFollowTask(index), null);
+                    persistentTasks.addTask(Integer.toString(i), ShardFollowTask.NAME,
+                        createShardFollowTask(new Index(index, IndexMetaData.INDEX_UUID_NA_VALUE)), null);
                 }
             }
             mdBuilder.put(imdBuilder);

--- a/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/TransportFollowStatsActionTests.java
+++ b/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/TransportFollowStatsActionTests.java
@@ -5,12 +5,16 @@
  */
 package org.elasticsearch.xpack.ccr.action;
 
+import org.elasticsearch.Version;
 import org.elasticsearch.cluster.ClusterName;
 import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.ByteSizeUnit;
 import org.elasticsearch.common.unit.ByteSizeValue;
 import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.Index;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.persistent.PersistentTasksCustomMetaData;
 import org.elasticsearch.test.ESTestCase;
@@ -24,30 +28,48 @@ import static org.hamcrest.Matchers.is;
 public class TransportFollowStatsActionTests extends ESTestCase {
 
     public void testFindFollowerIndicesFromShardFollowTasks() {
+        Settings indexSettings = Settings.builder()
+            .put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
+            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)
+            .build();
+
+        IndexMetaData index1 = IndexMetaData.builder("index1").settings(indexSettings).build();
+        IndexMetaData index2 = IndexMetaData.builder("index2").settings(indexSettings).build();
+        IndexMetaData index3 = IndexMetaData.builder("index3").settings(indexSettings).build();
+
         PersistentTasksCustomMetaData.Builder persistentTasks = PersistentTasksCustomMetaData.builder()
-            .addTask("1", ShardFollowTask.NAME, createShardFollowTask("abc"), null)
-            .addTask("2", ShardFollowTask.NAME, createShardFollowTask("def"), null);
+            .addTask("1", ShardFollowTask.NAME, createShardFollowTask(index1.getIndex()), null)
+            .addTask("2", ShardFollowTask.NAME, createShardFollowTask(index2.getIndex()), null)
+            .addTask("3", ShardFollowTask.NAME, createShardFollowTask(index3.getIndex()), null);
 
         ClusterState clusterState = ClusterState.builder(new ClusterName("_cluster"))
-            .metaData(MetaData.builder().putCustom(PersistentTasksCustomMetaData.TYPE, persistentTasks.build()).build())
+            .metaData(MetaData.builder()
+                .putCustom(PersistentTasksCustomMetaData.TYPE, persistentTasks.build())
+                // only add index1 and index2
+                .put(index1, false)
+                .put(index2, false)
+                .build())
             .build();
         Set<String> result = TransportFollowStatsAction.findFollowerIndicesFromShardFollowTasks(clusterState, null);
         assertThat(result.size(), equalTo(2));
-        assertThat(result.contains("abc"), is(true));
-        assertThat(result.contains("def"), is(true));
+        assertThat(result.contains(index1.getIndex().getName()), is(true));
+        assertThat(result.contains(index2.getIndex().getName()), is(true));
 
-        result = TransportFollowStatsAction.findFollowerIndicesFromShardFollowTasks(clusterState, new String[]{"def"});
+        result = TransportFollowStatsAction.findFollowerIndicesFromShardFollowTasks(clusterState,
+            new String[]{index2.getIndex().getName()});
         assertThat(result.size(), equalTo(1));
-        assertThat(result.contains("def"), is(true));
+        assertThat(result.contains(index2.getIndex().getName()), is(true));
 
-        result = TransportFollowStatsAction.findFollowerIndicesFromShardFollowTasks(clusterState, new String[]{"ghi"});
+        result = TransportFollowStatsAction.findFollowerIndicesFromShardFollowTasks(clusterState,
+            new String[]{index3.getIndex().getName()});
         assertThat(result.size(), equalTo(0));
     }
 
-    static ShardFollowTask createShardFollowTask(String followerIndex) {
+    static ShardFollowTask createShardFollowTask(Index followerIndex) {
         return new ShardFollowTask(
             null,
-            new ShardId(followerIndex, "", 0),
+            new ShardId(followerIndex, 0),
             new ShardId("leader_index", "", 0),
             1024,
             1024,

